### PR TITLE
Feature/react messenger

### PR DESCRIPTION
**Pull request by albangr**
- Redux setup
- Router setup
- Basic css generated just to have something to test, but everything is wip

- Messenger components : 
-> messenger
  -> chatList
  -> message
  -> messageForm
  -> messageList

I moved the chat list selection out of redux state to the router. So that we can use Messenger component as a standalone without the chatList component. 

i will build a container for Need to manage routing between different chats (providers).

### CODE REVIEW

File: frontend/actions/index.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  const promise = fetch(url).then(r => r.json());
+
+  return {
+    type: FETCH_MESSAGES,
+    payload: promise
+  };
+}
+
+export function createMessage(chat, author, content) {
+  const url = `${BASE_URL}/${chat}/messages`;
```
Comment (maximedelpit) [inline frontend/actions/index.js]: Same here, should call ActionCable action ?

File: frontend/components/messenger/messenger.css [hunk 2]
```diff
@@ -0,0 +1,136 @@
+.messaging-wrapper {
```
Comment (maximedelpit) [inline frontend/components/messenger/messenger.css]: Ideally should be split between components what do you think ?

File: frontend/containers/message_list/message_list.jsx [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import Message from '../../components/message/message';
+import MessageForm from '../../containers/message_form/message_form';
+
+class MessageList extends Component {
+  componentWillMount() {
+    this.fetchMessages();
+  }
+
+  componentDidMount() {
+    this.refresher = setInterval(this.fetchMessages, 5000);
```
Comment (maximedelpit) [inline frontend/containers/message_list/message_list.jsx]: This should be handled by ActionCable

File: frontend/actions/index.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+const BASE_URL = "https://wagon-chat.herokuapp.com";
+
+export const FETCH_MESSAGES = "FETCH_MESSAGES";
+export const MESSAGE_POSTED = "MESSAGE_POSTED";
+
+export function fetchMessages(chat) {
+  const url = `${BASE_URL}/${chat}/messages`;
```
Comment (maximedelpit) [inline frontend/actions/index.js]: More on my part but it should now come from ActionCable ? What do you thing ?

File: frontend/actions/index.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  return {
+    type: FETCH_MESSAGES,
+    payload: promise
+  };
+}
+
+export function createMessage(chat, author, content) {
+  const url = `${BASE_URL}/${chat}/messages`;
+  const body = { author, content };
```
Comment (maximedelpit) [inline frontend/actions/index.js]: Need to adapt to our DB params

File: frontend/reducers/messages_reducer.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { FETCH_MESSAGES, MESSAGE_POSTED } from "../actions";
+
+export default function(state = null, action) {
+  switch (action.type) {
+    case FETCH_MESSAGES: {
+      return action.payload.messages;
+    }
+    case MESSAGE_POSTED: {
+      const copiedState = state.slice(0);
```
Comment (maximedelpit) [inline frontend/reducers/messages_reducer.js]: I don't get why you slice this

File: frontend/components/messenger/messenger.css (hunk 2 above)
Comment (albangr) [reply frontend/components/messenger/messenger.css]: yes, as i said, i just put some css to render something on the view (to test my code). But this need to be redo & refacto

File: frontend/containers/message_list/message_list.jsx (hunk 3 above)
Comment (albangr) [reply frontend/containers/message_list/message_list.jsx]: yes we'll make a new action refresher that will handle ActionCable push

File: frontend/reducers/messages_reducer.js (hunk 6 above)
Comment (albangr) [reply frontend/reducers/messages_reducer.js]: I wanted to clone state to push the new message in it.
state.slice allow me to create a new array

File: frontend/actions/index.js (hunk 5 above)
Comment (albangr) [reply frontend/actions/index.js]: agreed, i just kept lewagon struct to test it with their API. I'll change that as soon as our Api works.

File: frontend/actions/index.js (hunk 4 above)
Comment (albangr) [reply frontend/actions/index.js]: maybe not this one, maybe we just manage an action addMessage who add the new external message to the redux state when he is created in db.

i'm not sure we want to refetch all messages in that case.

File: frontend/actions/index.js (hunk 1 above)
Comment (albangr) [reply frontend/actions/index.js]: maybe not here, this action will save something in db as soon as the API works. I think your cable needs to be trigger after_create of the message no ?

Comment (maximedelpit) [reply frontend/actions/index.js]: Well ActionCable work both way so you can call js_action that trigger rails methods in your channel. I propose to implement something and then check if consistent

File: frontend/reducers/messages_reducer.js (hunk 6 above)
Comment (maximedelpit) [reply frontend/reducers/messages_reducer.js]: Ok I missed that slice(0) just copy the whole array.

File: frontend/actions/index.js (hunk 4 above)
Comment (maximedelpit) [reply frontend/actions/index.js]: Makes sense

File: frontend/containers/message_list/message_list.jsx (hunk 3 above)
Comment (maximedelpit) [reply frontend/containers/message_list/message_list.jsx]: In my opinion the refresher should come from a lifecycle action (didUpdate / shouldComponentUpdate  / WillReceiveProp)

File: frontend/packs/messenger.jsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import messagesReducer from '../reducers/messages_reducer';
+
+import { BrowserRouter as Router, Route, Redirect, Switch } from 'react-router-dom';
+import { createBrowserHistory as history } from 'history';
+
+const identityReducer = (state = null) => state;
+
+const initialState = {
+  messages: [],
+  chats: ['general', 'react', 'paris'],
```
Comment (maximedelpit) [inline frontend/packs/messenger.jsx]: should be part of redux state ? or data-attributes initial state?

File: frontend/packs/messenger.jsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  currentUser: prompt("What is your username?") || `anonymous${Math.floor(10 + (Math.random() * 90))}`
```
Comment (maximedelpit) [inline frontend/packs/messenger.jsx]: should be part of redux state + no prompt 
 or data-attributes initial state?

File: frontend/actions/index.js (hunk 1 above)
Comment (maximedelpit) [reply frontend/actions/index.js]: the cable needs to be linked to the component when message received

**Pull request merged**